Packages: Add a basic Jetpack Logo package

tyxla opened the pull request:
This PR is an experiment to create and use a Jetpack Logo as a composer package. Similar to #12363, but I wanted to start it from scratch to go through the process, plus I introduced several changes, some of which are opinionated.

This is an initial attempt to address 3879-gh-jpop-issues.

I'm planning to use this PR for a base of several other experiments too. Feedback is appreciated.

**Note on design review**: The only design feedback I'm looking here is the alt attribute change we're doing here, in favor of better accessibility. See [convo](https://github.com/Automattic/jetpack/pull/12379#discussion_r286237047) for details.

#### Changes proposed in this Pull Request:

* Same as #12363, but has several more changes:
* **Design changes:** Changed the original `alt` attribute to just `Jetpack.` - see [convo](https://github.com/Automattic/jetpack/pull/12379#discussion_r286237047)
* Source directory renamed to `/src` instead of `/lib`, because it's more common for PHP libraries.
* Package dir renamed to `/logo` instead of `/jetpack-logo`, because `jetpack` is assumed and therefore arbitrary.
* Removed version from the namespace, because I think it can cause more problems than it would actually solve (also see this discussion: https://github.com/Automattic/jetpack/pull/12213#discussion_r281981451).
* Rename `utils` in the namespace to `assets`, because it's more specific.
* Rename class to just `Logo` - Jetpack already exists everywhere, including in the namespace.
* Fixed the license - now it's the same as Jetpack, the plugin.
* Refactored a bit to make easier to test and more flexible.
* Move the actual logo file to the package.
* Allow for custom logo URL in the class interface.
* Add a README with example usage.
* Ignore composer.lock in packages
* Add packages directory to the phpcs whitelist for linting.
* Add some temporary phpcs ignores - in the long term, we need to make PSR-4 and WP coding standards compatible.
* Added basic tests for the new package.
* Add `composer install` to the Travis CI setup.
* **Bump the minimum PHP version to 5.3.2** because of Composer's minimum requirements.
* Add a `php-build` script that basically runs `composer install` and add it to our build scripts.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
See p1HpG7-6V2-p2 and p1HpG7-6Ym-p2 for details.

#### Testing instructions:
* Make sure you're not connected to WP.com, but Jetpack is activated.
* Go to /wp-admin/plugins.php
* Make sure the logo is displayed properly.
* Run the tests and make sure they pass: `yarn docker:phpunit --testsuite packages`

#### Proposed changelog entry for your changes:
* Introduced a new Jetpack Logo package.

#### Notes
* Architecture of how tests are structured is likely going to change soon. While they're still in the global directory, we're going to move them out to the packages and make them run both independently and together as a whole. This should be good for now though.
* We'll have to build a fresh `composer.lock` after releasing the logo as a package.

tyxla committed:
Packages: Introduce a basic Logo composer package

tyxla committed:
Composer: Require our new Logo package

tyxla committed:
Core: Autoload

tyxla committed:
Connection Banner: Use Logo package

tyxla committed:
Temp: Disable some phpcs rules

tyxla committed:
Adds some docs, fix phpcs errors

tyxla committed:
Fix comment

tyxla committed:
Allow custom URL in static rendering

tyxla committed:
Add README

tyxla committed:
Ignore composer.lock in packages

tyxla committed:
Remove logo from the old location

tyxla committed:
Add a testsuite for packages

tyxla committed:
Add some Logo package tests

tyxla committed:
Composer update

jeherve commented on .phpcs.xml.dist:
```diff
@@ -1,10 +1,15 @@
 	<rule ref="PHPCompatibilityWP"/>
-	<rule ref="WordPress-Core" />
+	<rule ref="WordPress-Core">
+		<!-- Can we disable this just for `packages/*`? -->
```
I think it may be possible:

```
<!--
    You can hard-code ignore patterns for specific sniffs,
    a feature not available on the command line. Please note that
    all sniff-specific ignore patterns are checked using absolute paths.

    The code here will hide all messages from the Squiz DoubleQuoteUsage
    sniff for files that match either of the two exclude patterns.
 -->
 <rule ref="Squiz.Strings.DoubleQuoteUsage">
    <exclude-pattern>*/tests/*</exclude-pattern>
    <exclude-pattern>*/data/*</exclude-pattern>
 </rule>
```
-- https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset

tyxla committed:
Composer install for Travis

tyxla committed:
Bump PHP to 5.6

tyxla committed:
Disable PSR-4 conflicting sniffs only for packages

tyxla committed:
Add packages to phpcs whitelist

gravityrail commented:
> Bump to PHP 5.6, which we'll be able to do without remorse when the new WP major version is released.

We have to support the previous couple of versions of WordPress, not just the latest version. We cannot depend on 5.6 support for ~6 months from now.

It should be fine to bump to PHP 5.3 for now, as it includes all the things we need (right?)

gititon commented:
https://github.com/Automattic/jetpack/pull/new/try/composer-jetpack-logo-package-singleton-interfaces

We got it to work, implemented changes to make it a full fledged singleton, added a common interfaces package for our packages to use, made the logo package implement it. 

We explored setting the common interfaces package as a dependency of the logo package instead of including it via top level composer.json, but haven't cracked that yet.

tyxla committed:
Support PHP 5.3

tyxla committed:
Cleanup travis include matrix

tyxla committed:
Cleanup setup-travis.sh

tyxla commented:
@gravityrail FWIW there hasn't been a PHP 5.2 to 5.3 minimum requirement bump in the WP core. It was directly bumped to 5.6. Also, I've removed the changes from this PR and rebased to the latest master, and it seems that since #12287 and #12395 we're no longer running tests against 5.3, so this is not something that this PR introduces.

simison commented on packages/.gitignore:
```diff
@@ -0,0 +1,2 @@
+## Ignore composer.lock for our library packages
+/**/composer.lock
```
Can this still get generated and can the untracked presence of this file produce different results for different people?

Is there a way to disable it being generated in the first place?

tyxla replied:
> Is there a way to disable it being generated in the first place?

Not to my knowledge.

> Can the untracked presence of this file produce different results for different people?

It can, but that's generally the idea for libraries - they should be flexible and reusable in different contexts. As far as I understand they are expected to produce different results, and that should be just fine. If it's not, then it's highly possible that the library is doing something wrong.

simison replied:
Valid point about package dynamics!

Do these dependencies still get locked in Jetpack repo in the root lockfile, then?

If they do, great!

If not, wouldn't we have a situation where Jetpack's dependencies are not consistent for everyone in the monorepo/plugin between developers and the CI? Each time people type `composer install`, they'd get different versions. That's not the case for consumers of packaged versions of these monorepo-packages, who have their own lockfiles and need to type `composer update` to bump versions of transitive deps in packages anyway.

Could the lockfile be stripped out when mirroring these packages to their own repositories? Lockfile could even just be left in and when packages are used as a dependency, the lockfile will be ignored:

> _For your library you may commit the composer.lock file if you want to. This can help your team to always test against the same dependency versions. However, this lock file will not have any effect on other projects that depend on it. It only has an effect on the main project._

([via](https://getcomposer.org/doc/02-libraries.md#lock-file))

There might be angle I'm missing tho but just wanted to make sure deps stay locked in main repo as they should. :-)

Sidenote: this just reminded me of this situation in the past where the main repo didn't have package-lock commited https://github.com/Automattic/newspack-blocks/pull/6#issuecomment-482074977

tyxla replied:
> Do these dependencies still get locked in Jetpack repo in the root lockfile, then?

Yes. The idea is we should have `composer.lock` in the root of Jetpack (as it's a project), but not in the packages (which are libraries).

> but just wanted to make sure deps stay locked in main repo as they should

Yeah, that's understandable, and thanks for checking.  As you can see in this PR, we're committing the root `composer.lock` and we're only ignoring the ones in `packages/*`.

Let me know if you have any other concerns, though!

simison replied:
So installed dependency versions from `/packages/*/composer.json` files _do_ get hoisted to the root lockfile? Just want to make sure I understood you correctly.

tyxla replied:
I see where that gets confusing. 

It really depends on how we decide to consume the packages later inside Jetpack. But yeah, I think you're right, we eventually might have to end up adding the version specifically to `composer.json` of the packages if we want to consume the package as a symlink, and have a `composer.lock` inside Jetpack that's accurate about the package versions.

simison replied:
To elaborate little more; I'd expect Jetpack depending on its own `/packages/**` be fluid but any transitive 3rd party dependencies in `/packages/*/composer.json` files to get locked down in some lockfile that's version controlled, to avoid surprises.

(So exactly like in Calypso)

tyxla replied:
I see why that could be useful, but we still don't use any external libraries in our packages, and we might end up not doing that for a while. If we start doing it, and find the necessity, we can always start committing it. 

If folks feel strongly about this, we can commit it from the beginning - Composer's docs don't strictly forbid it.

tyxla replied:
To be fair though, without actually depending on any of the packages in an external from Jetpack project, we will never know if it's useful or not to have `composer.lock` committed for packages. I say let's stay open and sharp-eyed for it, and see what comes ahead.

aldavigdis commented on packages/logo/src/Logo.php:
```diff
@@ -0,0 +1,64 @@
+		return sprintf(
+			'<img src="%s" class="jetpack-logo" alt="%s" />',
+			esc_url( $this->url ),
+			esc_attr__(
```
If this is going to be a dependency of Jetpack, then we should look into and elimitate accessibility issues as we spot them.

This is a pretty good example, where the `alt` text provides *different* content as opposed to an actual alternative text to the Jetpack logotype.

aldavigdis commented on packages/logo/src/Logo.php:
```diff
@@ -0,0 +1,64 @@
+			'<img src="%s" class="jetpack-logo" alt="%s" />',
+			esc_url( $this->url ),
+			esc_attr__(
+				'Jetpack is a free plugin that utilizes powerful WordPress.com servers to enhance your site and simplify managing it',
```
```suggestion
				'Jetpack',
```

lezama replied:
Is there a different attribute we could use to put the old string as extra infortmation?

tyxla replied:
Yeah, I'd very much like keeping the original string somehow, while making everything accessible. Could we use an aria attribute to sort that out @aldavigdis?

aldavigdis replied:
I think this is more of a design issue. I suggest using a hyperlink wrapping the image tag somehow like this:

```html
<a
  href="https://www.jetpack.com/"
  rel="external"
  title="Visit the Jetpack website. Jetpack is a free plugin that utilizes powerful WordPress.com servers to enhance your site and simplify managing it."
  >
    <img src="foo.svg" alt="Jetpack." />
</a>
```

Notice that I added a reference to what the link does and a period after the end of the title and alt tags. This makes screen readers pause slightly.

tyxla replied:
Since I prefer not making the logo a link unless we want to, I've gone with the other suggestion - just changed the `alt` attribute to `Jetpack.`.

I'll ask for a design review because of this change too, just to be sure we're good with it.

tyxla replied:
Ah, I've also removed the localization function there, because we no longer need it - we don't translate Jetpack AFAIK.

aldavigdis replied:
I'd still wrap it with a `__` function, just in case.

tyxla replied:
I'm happy to do it if @Automattic/jetpack-crew are good with it - last time I did it, someone told me we aren't supposed to do it. Hence, I didn't do it here. 

If you ask me, it's unnecessary - we don't localize our plugin name, and that should be a good enough reason not to localize this string.

aldavigdis commented on packages/logo/src/Logo.php:
```diff
@@ -0,0 +1,64 @@
+			esc_url( $this->url ),
+			esc_attr__(
+				'Jetpack is a free plugin that utilizes powerful WordPress.com servers to enhance your site and simplify managing it',
+				'jetpack'
```
I have a hard time seeing where this `sprintf` attribute ends up in the `img` tag.

aldavigdis commented on packages/logo/src/Logo.php:
```diff
@@ -0,0 +1,64 @@
+	 */
+	public static function render( $url = '' ) {
+		$logo = new self( $url );
+
```
Just a nit-pick, but for a short render function like this, do we need this extra line break before the return statement for increased readability?

tyxla replied:
Not needed anymore, as the method has been removed.

aldavigdis commented on tests/php/packages/logo/test_Logo.php:
```diff
@@ -0,0 +1,42 @@
+	function test_render_img_tag() {
+		$output = Logo::render();
+		$url = home_url( '/wp-content/plugins/jetpack/packages/logo/assets/logo.svg' );
+		// Contains only a valid img tag.
```
Nit-pick: I'm generally a fan of newlines before comments.

tyxla replied:
Fixed in https://github.com/Automattic/jetpack/pull/12379/commits/1d36160f1651daeb7ec8406a5964a27b902f33b2.

tyxla committed:
Remove version from logo package composer.json

tyxla committed:
Remove any static methods

tyxla committed:
Fix usage in Jetpack_Connection_Banner

tyxla committed:
Fix tests

tyxla committed:
Add some newlines in tests

tyxla committed:
Update package README

tyxla committed:
Update composer.lock and composer.json

tyxla committed:
Shorten alt attribute

scottsweb commented:
> Design changes: Changed the original alt attribute to just Jetpack.

This makes a great deal of sense. We don't need such a long description in this context. 

If we are thinking purely about screen readers then it might be good to describe the logo, in this context though, using it as a heading to announce the content block works for me.

tyxla committed:
Bump the minimum PHP version because of Composer

tyxla committed:
Move autoloader below the PHP version check

tyxla committed:
Add dat dot.

lezama commented on jetpack.php:
```diff
@@ -133,6 +133,9 @@ function jetpack_admin_unsupported_php_notice() { ?>
 	return;
 }
 
+// Load all our packages.
```
could be other packages in the future too :)

tyxla replied:
Great point, corrected in https://github.com/Automattic/jetpack/pull/12379/commits/3d047e2c4dd4b41f03502499f7a3d31221250fac

tyxla committed:
Correct comment

aldavigdis commented:
> If we are thinking purely about screen readers then it might be good to describe the logo, in this context though, using it as a heading to announce the content block works for me.

I don't think we need to. We just need an alt text that represents the logotype.

tyxla committed:
Add a build-php step

lezama committed:
only ignore dev deps

enejb commented on packages/logo/src/Logo.php:
```diff
@@ -0,0 +1,47 @@
+	 */
+	public function __construct( $url = '' ) {
+		if ( ! $url ) {
+			$url = plugins_url( 'assets/logo.svg', __DIR__ );
```
Can we assume that each plugin that uses this library will have the logo in this place?

lezama replied:
it is relative to the file folder when you pass the second argument.

lezama replied:
and we are shipping `assets/logo.svg` with the package. 🤝

jeherve commented on composer.json:
```diff
@@ -18,9 +19,18 @@
 	},
 	"scripts": {
-		"php:compatibility": "composer install && vendor/bin/phpcs -p -s --runtime-set testVersion '5.2-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor --extensions=php",
+		"php:compatibility": "composer install && vendor/bin/phpcs -p -s --runtime-set testVersion '5.3-' --standard=PHPCompatibilityWP --ignore=docker,tools,tests,node_modules,vendor --extensions=php",
```
Good catch!

kraftbj replied:
For another PR, I would like to try this without the runtime. The version is set in .phpcs.xml already.

kraftbj commented on .travis.yml:
```diff
@@ -61,6 +61,7 @@ git:
 before_script:
   - export PLUGIN_SLUG=$(basename $(pwd))
   - export PATH="$HOME/.composer/vendor/bin:$PATH"
+  - composer install
```
Question: Should we have this here or https://github.com/Automattic/jetpack/blob/master/tests/setup-travis.sh#L16 ?

kraftbj commented on packages/logo/composer.json:
```diff
@@ -0,0 +1,12 @@
+{
+	"name": "automattic/jetpack-logo",
```
I think we need to add a version here for composer to install it until we have the packagist with a tag setup. I think `composer install` is failing for me because there's not a v1.

Once we're setup with packagist, it will automatically handling versioning based on the git tag.

kraftbj replied:
Disregard I think. Picked up the last rebase and seems to work now.

jeherve reviewed (commented):
I seem to be running into the same problem @kraft described here:
https://github.com/Automattic/jetpack/pull/12379#discussion_r286663896

To reproduce:
`rm -rf vendor`
`rm composer.lock`
`yarn build-production`

jeherve commented on jetpack.php:
```diff
@@ -133,6 +133,9 @@ function jetpack_admin_unsupported_php_notice() { ?>
 }
 
+// Load all the packages.
+require plugin_dir_path( __FILE__ ) . '/vendor/autoload.php';
```
Do you think we could add a safeguard here, to avoid fataling sites where `composer install` hasn't been run yet? Something like:

```
diff --git a/jetpack.php b/jetpack.php
index 0f450a7d9..59bd48868 100644
--- a/jetpack.php
+++ b/jetpack.php
@@ -118,6 +118,27 @@ function jetpack_admin_unsupported_php_notice() { ?>
 	<?php
 }
 
+/**
+ * Outputs an admin notice for folks running Jetpack without having run composer install.
+ *
+ * @since 7.4.0
+ */
+function jetpack_admin_missing_autoloader() { ?>
+	<div class="notice notice-error is-dismissible">
+		<p>
+		<?php
+		printf(
+			/* translators: Placeholder is a link to a support document. */
+			__( 'Your installation of Jetpack is incomplete. If you installed Jetpack from GitHub, please refer to <a href="%1$s" target="_blank" rel="noopener noreferrer">this document</a> to set up your development environment.', 'jetpack' ),
+			esc_url( 'https://github.com/Automattic/jetpack/blob/master/docs/development-environment.md' )
+		);
+		?>
+		</p>
+	</p>
+	</div>
+	<?php
+}
+
 if ( version_compare( phpversion(), JETPACK__MINIMUM_PHP_VERSION, '<' ) ) {
 	if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
 		error_log(
@@ -134,7 +155,23 @@ if ( version_compare( phpversion(), JETPACK__MINIMUM_PHP_VERSION, '<' ) ) {
 }
 
 // Load all the packages.
-require plugin_dir_path( __FILE__ ) . '/vendor/autoload.php';
+$jetpack_autoloader = JETPACK__PLUGIN_DIR . '/vendor/autoload.php';
+if ( is_readable( $jetpack_autoloader ) ) {
+	require $jetpack_autoloader;
+} else {
+	if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+		error_log(
+			sprintf(
+				/* translators: Placeholder is a link to a support document. */
+				__( 'Your installation of Jetpack is incomplete. If you installed Jetpack from GitHub, please refer to <a href="%1$s" target="_blank" rel="noopener noreferrer">this document</a> to set up your development environment.', 'jetpack' ),
+				esc_url( 'https://github.com/Automattic/jetpack/blob/master/docs/development-environment.md' )
+			)
+		);
+	}
+	add_action( 'admin_notices', 'jetpack_admin_missing_autoloader' );
+	return;
+}
+
 
 add_filter( 'jetpack_require_lib_dir', 'jetpack_require_lib_dir' );
 add_filter( 'jetpack_should_use_minified_assets', 'jetpack_should_use_minified_assets', 9 );

```

lezama replied:
🚢 IT

jeherve replied:
done in 500e6645a5a3fd22b526877fc086cfc30addea66

jeherve commented on jetpack.php:
```diff
@@ -133,6 +133,9 @@ function jetpack_admin_unsupported_php_notice() { ?>
 }
 
+// Load all the packages.
+require plugin_dir_path( __FILE__ ) . '/vendor/autoload.php';
```
Noting that this diff would solve that:
https://github.com/Automattic/jetpack/pull/12379#discussion_r286677626

jeherve replied:
done in 500e6645a5a3fd22b526877fc086cfc30addea66

jeherve committed:
Fail gracefully if the autoloader is not available yet.

See https://github.com/Automattic/jetpack/pull/12379#discussion_r286677626 for more discussion about this.

kraftbj commented:
Some more we can do to make it truly sing, but let's get this in as a public dress rehearsal.
